video: H.264 video format support

thecapn32 opened the pull request:
Add optional H.264 support to the video software generator.

The H.264 stream is generated at build time using ffmpeg and embedded as an
Annex-B byte stream. The generator splits the stream by AUD NAL units and
copies one access unit into each video buffer.

The feature is disabled by default and enabled with:

CONFIG_VIDEO_SW_GENERATOR_H264=y

Build-time options include width, height, FPS, number of frames, bitrate,
and H.264 profile.

josuah reviewed (commented):
Thanks for this improvement! Timely with the #92884 pull request doing the same thing but with real hardware, as well as every future video device.

Some quick feedback first, I will have to give a more in-depth round later.

josuah commented on drivers/video/video_h264_test.h:
This would also remove the amount of data on the repo.
Gstreamer and FFmpeg can generate files.

@JarmouniA Would it make sense to add these commands to CMake so that the user does not have to manually copy-paste a command in order to build? This would prevent to use it for i.e. CI.

It does not hurt to also document it though.

thecapn32 replied:
yes!
To generate the H.264 file
```sh
ffmpeg -f lavfi -i testsrc2=duration=5:size=640x480:rate=30 \
  -c:v h264 -b:v 500k -an test.h264
```
Convert it to C array:
```sh
xxd -i test.h264 > video_h264_test.h
```

josuah replied:
For later: an example of how a custom command might be integrated in CMakeLists.txt:

https://github.com/zephyrproject-rtos/zephyr/blob/0520dfe8ff931afb9f46648241538cbbc5ab0ea4/samples/subsys/llext/edk/ext1/CMakeLists.txt#L24-L33

It seems like `xxd` is currently accepted as a dependency already, we can use the same approach.

josuah commented on drivers/video/video_sw_generator.c:
```diff
@@ -215,6 +218,31 @@ static uint16_t video_sw_generator_fill_bayer8(uint8_t *buffer, uint16_t width,
+	static size_t position = 0;
+    size_t copy_size;
+
+	/* Calculate copy size */
+	size_t remaining = h264_test_data_len - position;
+	copy_size = (remaining > buffer_size) ? buffer_size : remaining;
```
It seems like I was mistaken: there are two valid approaches for `V4L2_PIX_FMT_H264`.

If I understand it right, what you propose is the stateful video encoder, where the data produced is immediately usable for being transferred (i.e. over the network).

> A stateful video encoder takes raw video frames in display order and encodes them into a bytestream. It generates complete chunks of the bytestream, including all metadata, headers, etc. The resulting bytestream does not require any further post-processing by the client. - https://www.kernel.org/doc/html/latest/userspace-api/media/v4l/dev-encoder.html

And it seems like it matches the description of `V4L2_PIX_FMT_H264` when `V4L2_FMT_FLAG_CONTINUOUS_BYTESTREAM` is enabled...

All good then! No need to change, but this opens the question of how to make the distinction bewteen "stateful" and "stateless" encoder in Zephyr.

hfruchet-st commented on include/zephyr/drivers/video.h:
```diff
@@ -1553,6 +1553,21 @@ int64_t video_get_csi_link_freq(const struct device *dev, uint8_t bpp, uint8_t l
+ */
+#define VIDEO_PIX_FMT_H264_NO_SC VIDEO_FOURCC('A', 'V', 'C', '1')
+
+/**
+ * H264 without start code
+ */
+#define VIDEO_PIX_FMT_H264_MVC VIDEO_FOURCC('M', '2', '6', '4')
+
```
In order to move forward on PR https://github.com/zephyrproject-rtos/zephyr/pull/92884, I've cherry-picked this commit but without the M264 definition.

josuah reviewed (commented):
Sorry for the long wait!

I think once the big `.h` is turned into a small line of ffmpeg this can be merged as a first early version that can be improved later.

thecapn32 commented:
sorry for the long delay but I still want to work on this

thecapn32 committed:
driver: video: sw_generator: Add H.264 support

Add optional H.264 test stream support to the video software generator.
Generate the Annex-B stream at build time with ffmpeg and copy one access
unit into each video buffer.

Signed-off-by: Mohammad Massoudi <m.massoudi32@gmail.com>

josuah commented on drivers/video/CMakeLists.txt:
```diff
@@ -62,6 +62,60 @@ if(CONFIG_VIDEO_SW_GENERATOR)
+            message(FATAL_ERROR
+                "CONFIG_VIDEO_SW_GENERATOR_H264 requires ffmpeg with h264_metadata bitstream filter"
+            )
+        endif()
```
There might be a lot of small details that go wrong such as the one you pointed, it could be possible to simplify this CMakeLists.txt by letting the FFmpeg command print the original error message and fail?

josuah commented on drivers/video/CMakeLists.txt:
```diff
@@ -62,6 +62,60 @@ if(CONFIG_VIDEO_SW_GENERATOR)
+                -f h264
+                ${H264_STREAM}
+            VERBATIM
+        )
```
With the goal of reducing the size, how about directly using the Kconfigs?

```suggestion
        string(
          CONCAT H264_INPUT
            "testsrc2=duration=${CONFIG_VIDEO_SW_GENERATOR_H264_DURATION}"
            ":size=${CONFIG_VIDEO_SW_GENERATOR_H264_WIDTH}"
            "x${CONFIG_VIDEO_SW_GENERATOR_H264_HEIGHT}"
            ":rate=${CONFIG_VIDEO_SW_GENERATOR_H264_FPS}"
        )
        
        file(MAKE_DIRECTORY ${ZEPHYR_BINARY_DIR}/include/generated)

        add_custom_command(
          OUTPUT
            ${ZEPHYR_BINARY_DIR}/include/generated/video_sw_generator_h264.h264
          COMMAND
            ffmpeg -y -f lavfi -i ${H264_INPUT} -f h264 -an
            -b:v ${CONFIG_VIDEO_SW_GENERATOR_H264_BITRATE_KBPS}k
            -bsf:v h264_metadata=aud=insert
            ${ZEPHYR_BINARY_DIR}/include/generated/video_sw_generator_h264.h264
          VERBATIM
        )
```

josuah replied:
There is an unwritten coding style for CMake, on top of the 2-space per indentation level:

- When folding long macro (such as here, `add_custom_comand(...)`, the keywords (like `OUTPUT`, `COMMAND`, `VERBATIM`) are on their own line
- They take parameters on the next line with extra indentation level.

The coding style says "100 lines *when possible*", so I guess we are all good.

But the way you did also seems fine, that is just about fine-tuning.

josuah commented on drivers/video/CMakeLists.txt:
```diff
@@ -62,6 +62,60 @@ if(CONFIG_VIDEO_SW_GENERATOR)
+            VERBATIM
+        )
+
+        generate_inc_file_for_target(
+            ${ZEPHYR_CURRENT_LIBRARY}
+            ${H264_STREAM}
+            ${ZEPHYR_BINARY_DIR}/include/generated/h264_frame_buffer.inc
+        )
```
Then will also need a change here...

```suggestion
        generate_inc_file_for_target(
          ${ZEPHYR_CURRENT_LIBRARY}
          ${ZEPHYR_BINARY_DIR}/include/generated/video_sw_generator_h264.h264
          ${ZEPHYR_BINARY_DIR}/include/generated/video_sw_generator_h264.inc
        )
```

How about using the same name `video_sw_generator_h264.inc` where only the extension changes?

Note that CMake uses [2 spaces for indentation](https://docs.zephyrproject.org/latest/contribute/style/cmake.html), this was a review mistake in previous commit.

josuah commented on drivers/video/Kconfig.sw_generator:
```diff
@@ -34,6 +34,58 @@ config VIDEO_SW_GENERATOR_PNG_HEIGHT
+	default 1
+	depends on VIDEO_SW_GENERATOR_H264
+	help
+	  Duration in seconds of the generated H.264 test stream.
```
1 second at 30 FPS is a lot of frames already, and not possible to go below 1 if it is in seconds. How about a number of frames for constrained platforms?

For instance, `VIDEO_SW_GENERATOR_H264_NUM_FRAMES`.

josuah commented on drivers/video/Kconfig.sw_generator:
```diff
@@ -34,6 +34,58 @@ config VIDEO_SW_GENERATOR_PNG_HEIGHT
 
+config VIDEO_SW_GENERATOR_H264
+	bool "H.264 stream generated by Video Software Generator"
+	default n
```
`default n` is already the default, so you can remove it

josuah commented on drivers/video/Kconfig.sw_generator:
```diff
@@ -34,6 +34,58 @@ config VIDEO_SW_GENERATOR_PNG_HEIGHT
+config VIDEO_SW_GENERATOR_H264_WIDTH
+	int "The width of the H.264 frame generated by Video Software Generator"
+	default 320
+	depends on VIDEO_SW_GENERATOR_H264
```
Rather than a `depends on` for every line, I would encourage an `if ...`/`endif` block around all the related options

josuah commented on drivers/video/Kconfig.sw_generator:
```diff
@@ -34,6 +34,58 @@ config VIDEO_SW_GENERATOR_PNG_HEIGHT
+	help
+	  ffmpeg encoder name used to generate the H.264 test stream. Depending
+	  on the ffmpeg build host configuration, users may need to select a
+	  specific encoder such as libx264 or libopenh264.
```
Not specify any value would allow FFmpeg to select the default for the particular platform? If so that is going to help with compatibility with MacOS + Windows.

This is also fewer options to maintain.

josuah reviewed (commented):
Thank you for this update! Some small adjustments proposed here regarding the configuration.

josuah commented on drivers/video/video_sw_generator.c:
```diff
@@ -110,9 +120,106 @@ static const struct video_format_cap fmts[] = {
+	return (h264_frame_buffer[start + prefix_len] & 0x1f) == 9;
+}
+
+static bool video_sw_generator_h264_next_access_unit(size_t offset, size_t *start, size_t *end)
```
Zephyr uses `int` as return types. 0 == ok, -ERRORNUMBER (such as `-EIO`) == error

josuah commented on drivers/video/video_sw_generator.c:
```diff
@@ -110,9 +120,106 @@ static const struct video_format_cap fmts[] = {
+		    h264_frame_buffer[i + 2] == 0x01) {
+			*start = i;
+			*prefix_len = 3;
+			return true;
```
For all the hardcoded numbers, we would need some matching `#define`s that tell what they are

josuah commented on drivers/video/video_sw_generator.c:
```diff
@@ -110,9 +120,106 @@ static const struct video_format_cap fmts[] = {
+	return false;
+}
+
+static bool video_sw_generator_h264_is_aud(size_t start, size_t prefix_len)
```
Is it for checking that the frames are audio-only? Not needed as the source of data is well known at compile time

thecapn32 replied:
No, Access Unit Delimiter(AUD) is an H.264 NAL unit type, and the driver uses it as the boundary/start marker for each generated access unit, effectively one encoded frame in this generated stream.

josuah replied:
Yes thank you that makes sense, good to have confirmation.

josuah commented on drivers/video/video_sw_generator.c:
```diff
@@ -302,6 +419,35 @@ static uint16_t video_sw_generator_fill_bayer8(uint8_t *buffer, uint16_t width,
+	frame_size = end - start;
+	if (vbuf->size < frame_size) {
+		LOG_ERR("Buffer too small (need %zu, have %zu)", frame_size, vbuf->size);
+		return -ENOMEM;
```
`-ENOBUFS` would help distinguishing a "memory allocation" error from a "not enough buffer space" error

josuah reviewed (commented):
And here some comments about the content of the driver itself.

josuah commented on drivers/video/video_sw_generator.c:
```diff
@@ -110,9 +126,121 @@ static const struct video_format_cap fmts[] = {
+	return -EINVAL;
+}
+
+static bool video_sw_generator_h264_is_aud(size_t start, size_t prefix_len)
```
A small comment will give a lot of insights:

`/* Verify if the current NAL block at position "start" is a AUD (access unit delmiter) */`

<img width="845" height="698" alt="Image" src="https://github.com/user-attachments/assets/d914baaf-1d81-4f42-99b6-2241ace845bd" />

(^ just to show that this is the first block of a sequence of NAL)

josuah commented on drivers/video/video_sw_generator.c:
```diff
@@ -110,9 +126,121 @@ static const struct video_format_cap fmts[] = {
+	*end = sizeof(h264_frame_buffer);
+	next = *start + prefix_len;
+
+	while (video_sw_generator_h264_start_code(next, &next, &next_prefix_len) == 0) {
```
A small comment will help people new to H.264 (like me!) to quickly grasp what the standard is and how it is used:

`/* Skip NALs blocks until the next AUD (access unit delimiter) */`

josuah commented on drivers/video/video_sw_generator.c:
```diff
@@ -110,9 +126,121 @@ static const struct video_format_cap fmts[] = {
+	return true;
+}
+
+static int video_sw_generator_h264_start_code(size_t offset, size_t *start, size_t *prefix_len)
```
A small comment will help giving a quick insight about how start code work:

`/* Scan through raw bytes of the buffer searching for the next start code */`

josuah replied:
Maybe calling it `_h264_next_start_code` instead of `_h264_start_code` will help further.

josuah reviewed (commented):
Thank you once again for this feature! That sounds robust, I think this is getting ready.

I believe it will help to add some hints for people who are completely new to H.264 and would want to try it using `video-sw-generator`.

josuah reviewed (dismissed):
All looking good thanks!

josuah commented:
> dynamic way to adjust video format in video-sw-generator ( now the H.264 format is static)

I am not sure this is possible without re-encoding the feed, which would need to write a complete encoder, a bit large for the scope of a test pattern generator.

> add mechanism to send h264 frames to maintain 30FPS

Good work with the frame parser! Now it is sending frames accurately at the FPS configured.

ngphibang commented on drivers/video/CMakeLists.txt:
```diff
@@ -62,6 +62,31 @@ if(CONFIG_VIDEO_SW_GENERATOR)
+        set(H264_INPUT
+            "testsrc2=size=${CONFIG_VIDEO_SW_GENERATOR_H264_WIDTH}x${CONFIG_VIDEO_SW_GENERATOR_H264_HEIGHT}"
+        )
+        string(APPEND H264_INPUT ":rate=${CONFIG_VIDEO_SW_GENERATOR_H264_FPS}")
```
I think we should add a `find_program()`
first to ensure there is ffmpeg installed on the host (and where it is) then message a compilation logs it it is not installed.

josuah replied:
Like it was initially. Sorry @thecapn32 I indicated you the wrong way.

ngphibang commented on drivers/video/CMakeLists.txt:
```diff
@@ -62,6 +62,31 @@ if(CONFIG_VIDEO_SW_GENERATOR)
+
+        add_custom_command(
+            OUTPUT
+              ${ZEPHYR_BINARY_DIR}/include/generated/video_sw_generator_h264.h264
+            COMMAND
+              ffmpeg -y -f lavfi -i ${H264_INPUT} -f h264 -an
+              -b:v ${CONFIG_VIDEO_SW_GENERATOR_H264_BITRATE_KBPS}k
+              -frames:v ${CONFIG_VIDEO_SW_GENERATOR_H264_NUM_FRAMES}
+              -bsf:v h264_metadata=aud=insert
+              ${ZEPHYR_BINARY_DIR}/include/generated/video_sw_generator_h264.h264
```
Should we add some more critical parameters for H.264, for example, at least  the `profile` ?

josuah replied:
I think @thecapn32 was interested in introducing a video control for this.

This could be some way to add support for it, a read-only control for which kind of profile was used.

Then it is possible to integrate this control in various APIs like libMP.

Not possible to switch but at least make use of the information.

ngphibang replied:
Support of JPEG and H.264 in video_sw_generator is like a quick way to have something to test without libMP as these supports simplify the real-world scenario (with the "parser" included inside the emulator itself). In libMP, we stream the jpeg or h.264 directly from file and has a (jpeg/h264) parser to parse the bytestream before decoding it (https://github.com/zephyrproject-rtos/zephyr/pull/108348)

josuah replied:
video-sw-generator is useful as a test utility but I agree that `filesrc` will be some much more generic way.

thecapn32 replied:
I have added these paramters:
- profile with constrained_baseline, baseline, normal and high options
- pixel format fixed to yuv420p
- level fixed to 3.0

Let me know if I am missing anything.

ngphibang commented on drivers/video/video_sw_generator.c:
```diff
@@ -302,6 +443,35 @@ static uint16_t video_sw_generator_fill_bayer8(uint8_t *buffer, uint16_t width,
+
+	frame_size = end - start;
+	if (vbuf->size < frame_size) {
+		LOG_ERR("Buffer too small (need %zu, have %zu)", frame_size, vbuf->size);
```
`vbuf->size` is `uint32_t` so it should be `%u`

avolmat-st reviewed (approved):
Thanks @thecapn32. LGTM.

Tested it on the STM32N6-DK board via USB UVC

``west build -p -b stm32n6570_dk//sb --snippet video-sw-generator samples/subsys/usb/uvc/ -DCONFIG_VIDEO_SW_GENERATOR_H264=y -DCONFIG_VIDEO_SW_GENERATOR_H264_NUM_FRAMES=10``

then playing on the laptop

``gst-launch-1.0 v4l2src device=/dev/video4 ! video/x-h264 ! h264parse ! avdec_h264 ! autovideosink``
